### PR TITLE
Move some enums that are part of the public API.

### PR DESCRIPTION
**Pull request by jsha**
Move ProtocolVersion, CipherSuite, and SignatureScheme from src/msgs/enums.rs to src/enums.rs. This removes them from the `internals` part of the public API and preserves their existing visibility as `rustls::ProtocolVersion`, `rustls::CipherSuite`, and `rustls::SignatureScheme`. Doing this ensures that they are visible in the docs under only one name: the stable API one.

Part of #1036

### CODE REVIEW

File: rustls/src/lib.rs [hunk 1]
```diff
@@ -336,6 +336,9 @@ mod x509;
 mod check;
 mod bs_debug;
 mod builder;
+#[allow(non_camel_case_types)]
```
Comment (djc) [inline rustls/src/lib.rs:339]: If we need these, I think they should go in an inner attribute inside the module. Why do we need them?

File: rustls/src/lib.rs [hunk 2]
```diff
@@ -383,6 +383,7 @@ pub use crate::ticketer::Ticketer;
 pub use crate::tls12::Tls12CipherSuite;
 pub use crate::tls13::Tls13CipherSuite;
 pub use crate::versions::{SupportedProtocolVersion, ALL_VERSIONS, DEFAULT_VERSIONS};
+pub use enums::*;
```
Comment (djc) [inline rustls/src/lib.rs:386]: I prefer to never use `*` imports, as they make it less obvious where names come from.

File: rustls/src/msgs/enums_test.rs [hunk 3]
```diff
@@ -15,7 +15,7 @@ fn get16<T: Codec>(enum_value: &T) -> u16 {
     (enc[0] as u16 >> 8) | (enc[1] as u16)
 }
 
-fn test_enum16<T: Codec>(first: T, last: T) {
+pub(crate) fn test_enum16<T: Codec>(first: T, last: T) {
```
Comment (djc) [inline rustls/src/msgs/enums_test.rs:18]: Maybe this should also be `#[cfg(test)]`?

File: rustls/src/msgs/handshake.rs [hunk 4]
```diff
@@ -1,15 +1,16 @@
-use crate::key;
+use crate::enums::{CipherSuite, ProtocolVersion, SignatureScheme};
 use crate::msgs::base::{Payload, PayloadU16, PayloadU24, PayloadU8};
 use crate::msgs::codec;
 use crate::msgs::codec::{Codec, Reader};
 use crate::msgs::enums::ECCurveType;
+use crate::msgs::enums::HandshakeType;
```
Comment (djc) [inline rustls/src/msgs/handshake.rs:6]: Style nit: I think it'd be better to concentrate `msgs::enums` imports in as few lines as possible.

File: rustls/src/msgs/message.rs [hunk 5]
```diff
@@ -1,11 +1,12 @@
+use crate::enums::ProtocolVersion;
 use crate::error::Error;
 use crate::msgs::alert::AlertMessagePayload;
 use crate::msgs::base::Payload;
 use crate::msgs::ccs::ChangeCipherSpecPayload;
 use crate::msgs::codec::{Codec, Reader};
+use crate::msgs::enums::ContentType;
```
Comment (djc) [inline rustls/src/msgs/message.rs:7]: Style nit: coalesce this with `HandshakeType` at least?

File: rustls/src/msgs/mod.rs [hunk 6]
```diff
@@ -12,6 +12,7 @@ pub mod deframer;
 pub mod enums;
 pub mod fragmenter;
 #[allow(non_camel_case_types)]
+#[macro_use]
```
Comment (djc) [inline rustls/src/msgs/mod.rs:15]: I think it's better to specifically import the required macros by name, which I think is the idiomatic style starting with 2018 edition.

File: rustls/src/server/hs.rs [hunk 7]
```diff
@@ -206,7 +205,7 @@ impl ExtensionProcessing {
             .is_some()
             || hello
                 .cipher_suites
-                .contains(&CipherSuite::TLS_EMPTY_RENEGOTIATION_INFO_SCSV);
+                .contains(&crate::enums::CipherSuite::TLS_EMPTY_RENEGOTIATION_INFO_SCSV);
```
Comment (djc) [inline rustls/src/server/hs.rs:208]: Style nit: this line is doing enough work as it is, probably better to import `CipherSuite` in some enclosing scope.

File: rustls/src/server/server_conn.rs [hunk 8]
```diff
@@ -1,14 +1,14 @@
 use crate::builder::{ConfigBuilder, WantsCipherSuites};
 use crate::conn::{CommonState, ConnectionCommon, Side, State};
+use crate::enums::ProtocolVersion;
```
Comment (djc) [inline rustls/src/server/server_conn.rs:3]: Nit: use the opportunity to fold these into a single line?

File: rustls/src/enums.rs [hunk 9]
```diff
@@ -0,0 +1,522 @@
+/// This file is autogenerated.  See https://github.com/ctz/tls-hacking/
```
Comment (djc) [inline rustls/src/enums.rs:1]: I think this is no longer correct. @ctz?

**Review by djc — COMMENTED** [review]
I think this is a good direction.

I added some comments on import style. IMO the style is a bit messy and I've been trying to clean it up over time when making changes that required different imports. I've been working into a direction where we have fewer lines of imports with repeated prefixes.

File: rustls/src/lib.rs (hunk 1 above)
Comment (jsha) [reply rustls/src/lib.rs:339]: I started down this path, and realized that the comments would be pointless repetition of the enum contents. For instance, commenting on `SSLv2` with "SSL Version 2" doesn't really help anybody. Similarly, commenting `TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256` with some prose explanation of its components would be pointless. The RFCs don't attempt to do that when registering the parameters.

**Comment by jsha** [comment]
Thanks to unpacking the `*` import, I realized I'd accidentally publicized the DecomposedSignatureScheme and SupportedSignatureSchemes types. I moved those back to their present position in `internals`.

I've kept the latest two commits separate for ease of review, but I'm happy to squash them into the first commit if you'd prefer.

File: rustls/src/lib.rs (hunk 1 above)
Comment (djc) [reply rustls/src/lib.rs:339]: Fair enough, but maybe the enums themselves should get a short docstring pointing to the RFC/IANA registry or whatever?

File: rustls/src/client/client_conn.rs [hunk 10]
```diff
@@ -1,14 +1,14 @@
 use crate::builder::{ConfigBuilder, WantsCipherSuites};
 use crate::conn::{CommonState, ConnectionCommon, Protocol, Side};
+use crate::enums::CipherSuite;
```
Comment (djc) [inline rustls/src/client/client_conn.rs:3]: So maybe fold these?

File: rustls/src/lib.rs (hunk 1 above)
Comment (djc) [reply rustls/src/lib.rs:339]: Would still be good to move these to inner attributes IMO.

File: rustls/src/msgs/handshake_test.rs [hunk 11]
```diff
@@ -2,6 +2,7 @@ use super::base::{Payload, PayloadU16, PayloadU24, PayloadU8};
 use super::codec::{put_u16, Codec, Reader};
 use super::enums::*;
 use super::handshake::*;
+use crate::enums::*;
```
Comment (djc) [inline rustls/src/msgs/handshake_test.rs:5]: Boy scout rule explicit exports (for this new line only)?

**Review by djc — COMMENTED** [review]
Almost there!

File: rustls/src/lib.rs (hunk 1 above)
Comment (jsha) [reply rustls/src/lib.rs:339]: I tried putting `#[allow(non_camel_case_types)]` at the top of rustls/src/enums.rs, but it did not appear to take effect for the declarations within the file.

I tried putting that attribute directly above the `enum_builder!` calls, but it similarly did not appear to take effect.

I tried putting that attribute inside the definition of `enum_builder!`, under the logic that most of the enums are non-camel case. But some (like HandshakeType) _are_ camel case, so I got "warning: unused attribute `allow`".

I'm sure there's a way to do this and I'll keep poking at it, but if you have an idea of how it should work, let me know.

Comment (djc) [reply rustls/src/lib.rs:339]: It needs to become an **inner** attribute, I think? `#![allow(non_camel_case_types)]`.

Comment (jsha) [reply rustls/src/lib.rs]: Ah, thanks for that. I hadn't grokked the difference between inner and outer attributes.

**Pull request merged**